Framework/PythonInterface/plugins/algorithms/DeltaPDF3D.py

```diff
@@ -0,0 +1,251 @@
+                    pass
+                logger.information('Using space group: '+space_group)
+                sg=SpaceGroupFactory.createSpaceGroup(space_group)
+            else:
+                check_space_group = False
+            for h in range(int(np.ceil(dimX.getMinimum())), int(np.floor(dimX.getMaximum()))+1):
+                for k in range(int(np.ceil(dimY.getMinimum())), int(np.floor(dimY.getMaximum()))+1):
+                    for l in range(int(np.ceil(dimZ.getMinimum())), int(np.floor(dimZ.getMaximum()))+1):
+                        if not check_space_group or sg.isAllowedReflection([h,k,l]):
+                            if cut_shape == 'cube':
+                                x_min=np.searchsorted(X,h-width[0])
+                                x_max=np.searchsorted(X,h+width[0])
+                                y_min=np.searchsorted(Y,k-width[1])
+                                y_max=np.searchsorted(Y,k+width[1])
+                                z_min=np.searchsorted(Z,l-width[2])
+                                z_max=np.searchsorted(Z,l+width[2])
+                                signal[x_min:x_max,y_min:y_max,z_min:z_max]=np.nan
+                            else:  # sphere
+                                signal[(Xs-h)**2/width[0]**2 + (Ys-k)**2/width[1]**2 + (Zs-l)**2/width[2]**2 < 1]=np.nan
+
+        if self.getProperty("CropSphere").value:
+            progress.report("Cropping to sphere")
+            sphereMin = self.getProperty("SphereMin").value
+            if sphereMin[0] < Property.EMPTY_DBL:
+                if len(sphereMin)==1:
+                    sphereMin = np.repeat(sphereMin, 3)
+                signal[Xs**2/sphereMin[0]**2 + Ys**2/sphereMin[1]**2 + Zs**2/sphereMin[2]**2 < 1]=np.nan
+            sphereMax = self.getProperty("SphereMax").value
+            if sphereMax[0] < Property.EMPTY_DBL:
+                if len(sphereMax)==1:
+                    sphereMax = np.repeat(sphereMax, 3)
+                signal[Xs**2/sphereMax[0]**2 + Ys**2/sphereMax[1]**2 + Zs**2/sphereMax[2]**2 > 1]=np.nan
+
+        if self.getProperty("Convolution").value:
+            progress.report("Convoluting signal")
+            signal = self._convolution(signal)
+
+        if self.getPropertyValue("IntermediateWorkspace"):
+            cloneWS_alg = self.createChildAlgorithm("CloneMDWorkspace", enableLogging=False)
+            cloneWS_alg.setProperty("InputWorkspace",inWS)
+            cloneWS_alg.execute()
+            signalOutWS = cloneWS_alg.getProperty("OutputWorkspace").value
+            signalOutWS.setSignalArray(signal)
+            self.setProperty("IntermediateWorkspace", signalOutWS)
+
+        # Do FFT
+        progress.report("Running FFT")
+        # Replace any remaining nan's or inf's with 0
+        # Otherwise you end up with a lot of nan's
+        signal[np.isnan(signal)]=0
+        signal[np.isinf(signal)]=0
+
+        signal=np.fft.fftshift(np.fft.fftn(np.fft.ifftshift(signal))).real
```

Comment:
I will try that out later but I don't really want to add a new dependency to Mantid just for this.

If it is faster I might optionally use pyFFTW over numpy.fft if it's available.